File: src/main/java/org/xyp/project/todoapp/user/Role.java
package org.xyp.project.todoapp.user;


import com.fasterxml.jackson.annotation.JsonFormat;
import jakarta.persistence.*;
import lombok.Data;
import org.xyp.project.todoapp.shared.enums.ActiveStatus;

@Data
@Entity
@Table(name = "todoapp_role")
public class Role {
    @EmbeddedId
    @JsonFormat(shape = JsonFormat.Shape.STRING)
    RoleId id;
    String name;
    @Enumerated(EnumType.STRING)
    ActiveStatus status;
    @Version
    Long optimisticVersion;
}

File: src/main/kotlin/org/xyp/todoapp/domain/user/Role.kt
package org.xyp.todoapp.domain.user

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonValue
import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import org.xyp.todoapp.core.enums.ActiveStatus
import java.time.LocalDateTime

@Entity
@Table(name = "todoapp_role")
class Role(

    @EmbeddedId
    @JsonFormat(shape = JsonFormat.Shape.STRING)
    val id: RoleId,

    var name: String? = null,

    @Enumerated(EnumType.STRING)
    var status: ActiveStatus? = null,

    @CreationTimestamp
    var createdAt: LocalDateTime? = null,
    @UpdateTimestamp
    var updatedAt: LocalDateTime? = null,

    @Version
    var optimisticVersion: Long? = null,
) {
    constructor(rid: RoleId) : this(id = rid, optimisticVersion = null)
}
@Embeddable
data class RoleId(
    @JsonValue
    val id: Long
)